src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -1840,19 +1932,51 @@ class NodeUtils(object):
         host_resc_avail = vnode_list[vnode_name].resources_available
         pbs.logmsg(pbs.EVENT_DEBUG4, '%s: host_resc_avail: %s' %
                    (caller_name(), host_resc_avail))
-        # Set the vnode type if supplied
-        if vntype:
+
+        # Set resources_available.vntype of natural node according
+        # to what's in local file if it needs to be propagated to server
+        if (vntype and "propagate_vntype_to_server" in self.cfg
+                and self.cfg['propagate_vntype_to_server']):
             host_resc_avail['vntype'] = vntype
-            pbs.logmsg(pbs.EVENT_DEBUG4, '%s: vnode type set to %s' %
-                       (caller_name(), vntype))
+            pbs.logmsg(pbs.EVENT_DEBUG4, '%s: vnode type set to %s'
+                       % (caller_name(), vntype))
+
         vnode_msg_cpu = '%s: vnode_list[%s].resources_available[ncpus] = %d'
         vnode_msg_mem = '%s: vnode_list[%s].resources_available[mem] = %s'
+
+        # Some kernel drivers "reserve" memory on /proc/meminfo
+        # and this is not reflected in the node meminfo; if necessary,
+        # adjust the values published as resources_available.mem
+        # on the per-socket vnodes by the difference spread over nodes
+        adjust_bytes_per_node = 0
+        if vnodes:
+            num_numa_nodes = len(self.numa_nodes)
+            total_nodemem = 0
+            for num in self.numa_nodes:
+                total_nodemem += size_as_int(self.numa_nodes[num]['MemTotal'])
+            total_hostmem = size_as_int(self.meminfo['MemTotal'])
+            pbs.logmsg(pbs.EVENT_DEBUG4,
+                       '%s: host memtotal %s; node memtotal %s'
+                       % (caller_name(),
+                          str(total_hostmem), str(total_nodemem)))
+            if total_hostmem < total_nodemem:
+                adjust_bytes_per_node = \
+                    int(math.ceil((total_nodemem-total_hostmem)
```

Comment:
Please add a space before and after the minus sign.

Reply:
Done